### PR TITLE
feat: add SQUAD_HOME env var and preset system

### LINKED ISSUE #1038: Feature: Global SQUAD_HOME for roaming agents across repos
### Problem

Today, squad agents only resolve from a local `.squad/` directory in the current repo. When I move between repos, I lose my squad. I have to either:
- Symlink `.squad/` into every repo (fragile)
- Re-init with snap-squad in every repo (copies, drift)
- Reference charters by absolute path (clunky)

### Proposal

Add a **global squad home** fallback so personal agents are always in scope, regardless of which repo I am in.

**Option A: Environment variable**
```bash
export SQUAD_HOME=~/src/dispatch/.squad
```
Squad CLI checks `SQUAD_HOME` when no local `.squad/` exists.

**Option B: Conventional global directory**
```
~/.squad/
  agents/
    triage/charter.md
    review/charter.md
```
Squad CLI checks `~/.squad/` as a fallback.

### Resolution order

1. **Repo-local** `.squad/` (highest priority — team/project-specific agents)
2. **Global** `SQUAD_HOME` or `~/.squad/` (personal agents that roam everywhere)

Local agents can override or extend global ones. If a repo has `.squad/agents/triage/charter.md`, it wins over the global triage agent.

### Why this matters

I work across many repos on multiple machines. I have a personal squad of agents I have tuned — triage, review, security, etc. I want them available everywhere without per-repo setup. This is the difference between:
- `npm install` (local to this project) → repo squad
- `npm install -g` (available everywhere) → global squad

### Integration with snap-squad

snap-squad presets would continue to scaffold repo-local `.squad/` directories. The global squad home is complementary — it is the personal baseline that snap-squad repos can extend or override.

### Context

This is part of a broader vision for a distributed nervous system architecture where agents, rules, and learnings roam across machines and repos. See: https://github.com/coreai-microsoft/dispatch-insiders/issues/74

---

## Incorporate Best of Snap Squad into Squad

Rather than keeping snap-squad as a separate tool, consider folding its best ideas directly into Squad:

### 1. Presets built into Squad CLI

```bash
squad init --preset default     # generalist squad
squad init --preset fast        # rapid POC, ship fast
squad init --preset specialists # deep expertise
squad init --preset my-squad    # custom personal preset
```

No need for a separate `npx snap-squad init`. Squad itself knows how to scaffold from presets. Users can publish and share custom presets.

### 2. Sub-agent routing

The real power of a squad is not just having agents — it is the dispatcher knowing which agent to route to. Squad should support:

- **Automatic routing** — based on the task, Squad picks the best agent (triage routes to triage agent, code review routes to reviewer, etc.)
- **Agent delegation** — one agent can hand off sub-tasks to another agent in the squad
- **Capability matching** — agents declare what they are good at, the router matches tasks to capabilities

```bash
# Instead of explicitly picking an agent:
squad run triage "review this PR"

# Squad figures it out:
squad run "review this PR"
# → routes to reviewer agent based on task analysis
```

### 3. The unified vision

| Today | Proposed |
|---|---|
| `snap-squad` scaffolds `.squad/` | `squad init --preset` does the same, built-in |
| Manual agent selection | Smart routing picks the best agent for the task |
| Agents are siloed | Agents can delegate sub-tasks to each other |
| Local to one repo | Global `SQUAD_HOME` + repo-local layering |

This makes Squad a complete system: scaffold, roam, route, delegate. One tool, not two.

### 4. Personal default preset

When I run `squad init` with no flags, it should use **my** default preset — not a generic one.

```bash
# First time: set my default
squad config set default-preset pauls-squad

# Now every repo I init just works:
squad init
# → uses pauls-squad preset automatically

# Override when needed:
squad init --preset fast
```

Where does the preset live? Options:

- **`~/.squad/presets/pauls-squad/`** — local preset directory, synced via dotfiles or git
- **GitHub repo** — `squad init --preset github:paulyuk/my-squad` pulls from a repo
- **`SQUAD_HOME` directory** — the global squad home IS the default preset. `squad init` copies from there.

The simplest version: `SQUAD_HOME` doubles as the default preset source. `squad init` with no args copies agents from `SQUAD_HOME` into the repo. No registry, no config file, just "use whatever I have globally."

```bash
export SQUAD_HOME=~/src/dispatch/.squad

cd ~/src/new-project
squad init
# → scaffolds .squad/ from ~/src/dispatch/.squad
# → Paul's agents are the default everywhere
```

### PR DESCRIPTION
**Pull request by paulyuk**
## Summary

Adds SQUAD_HOME env var support and a preset system for reusable agent configurations, with auto-init and `--remote` for roaming across machines.

> **Origin:** This feature is inspired by [snap-squad](https://github.com/paulyuk/snap-squad), which prototyped warm-start preset squads as an external addon. The plan is to retire snap-squad immediately and point users here instead — presets belong in squad itself.

### Recent Changes
- **`squad init --preset` auto-creates squad home** — if no presets directory exists, it initializes one automatically and seeds built-in presets. No setup step needed.
- **`squad preset init --remote`** — backs squad home with a private `squad-home` GitHub repo via `gh`. On a second machine, same command clones it.
- **`preset save` tips about `squad export`** for full squad snapshots (casting, skills, routing) — e.g. to publish to agent toolboxes.

---

## Quick Start

```bash
# Start a project with the built-in default preset
# (auto-creates ~/.squad/ and seeds presets if first run)
squad init --preset default

# Customize your agents — make it yours
# edit .squad/agents/*/charter.md, add/remove agents

# Save as your personal preset
squad preset save my-team

# Any new repo — one command
squad init --preset my-team
```

### Roam presets across machines

```bash
# Back your squad home with a private GitHub repo (one-time per machine)
squad preset init --remote

# On your second machine — clones your presets automatically
squad preset init --remote
squad init --preset my-team
```

> `squad preset init --remote` creates `<you>/squad-home` as a private repo via `gh repo create`, sets up `~/.squad/` as a clone, and pushes. On a second machine it detects the existing repo and clones it instead.

---

### SQUAD_HOME Environment Variable
A roaming squad root directory for user-level squad assets. Presets live here.

| Setup | Where presets live | Roaming? |
|-------|-------------------|----------|
| **Default** (auto-created on first `--preset` use) | `~/.squad/presets/` | ❌ Local only |
| **`--remote`** (recommended) | `~/.squad/presets/` backed by `<you>/squad-home` repo | ✅ Push/pull |
| **Custom `SQUAD_HOME`** | `$SQUAD_HOME/presets/` | Depends on backing |

**SDK functions:**
- `resolveSquadHome()` — resolves `SQUAD_HOME` env var or falls back to `~/.squad/`
- `ensureSquadHome()` — creates the directory if needed
- `resolvePresetsDir()` — returns the `presets/` subdirectory within SQUAD_HOME

### Preset System
Presets are named, reusable collections of agent charters stored in `SQUAD_HOME/presets/<name>/`.

**Built-in default preset** ships with 5 agents: lead, reviewer, devrel, security, docs.

**CLI Commands:**
```
squad init --preset <name>         # init + apply preset (auto-creates squad home if needed)
squad preset init --remote         # back squad home with private GitHub repo
squad preset init                  # local-only setup
squad preset list                  # list available presets
squad preset show <name>           # show preset details and agents
squad preset apply <name>          # apply preset to existing project
squad preset save <name>           # save current agents as a preset
```

**Collision policy:** existing agents are skipped by default; use `--force` to overwrite.

### Presets vs Export

| | **Presets** | **Export** |
|---|---|---|
| **What** | Agent charters only | Full squad snapshot (casting, skills, routing rules, history) |
| **Purpose** | Reusable starter kits for bootstrapping | Project portability, sharing configured squads, publishing to agent toolboxes |
| **Storage** | `SQUAD_HOME/presets/<name>/` | `squad-export.json` (portable file) |
| **Roaming** | Yes — via `--remote` or SQUAD_HOME backed by git/sync | Manual — share the export file |

### Testing
- 24 tests covering all preset functions including full round-trip (save → apply to new project)
- All existing tests continue to pass

### Changeset
Included — minor version bump for both `@bradygaster/squad-sdk` and `@bradygaster/squad-cli`.

Resolves #1038

### CODE REVIEW

**Review by tamirdresher — APPROVED** [review]
I'm good with this but you might want to consider the copilot comment about the env var pointing to something which is not a folder

**Comment by paulyuk** [comment]
### Update: preset save messaging + export distinction

**Changes in latest push:**

1. **`preset save` now clearly shows where presets are stored** — outputs the full path (e.g. `~/.squad/presets/my-team/`) so users always know where their preset landed.

2. **`preset save` now tips users about `squad export`** — after saving, the CLI prints:
   > *Tip: Presets save agents only (charters). For a full squad snapshot including casting state, skills, and routing rules — e.g. to share a configured squad or publish to an agent toolbox — use `squad export`.*

3. **PR description updated with Presets vs Export comparison** — table explaining when to use each:
   - **Presets** = reusable starter kits (agents/charters only, stored in SQUAD_HOME)
   - **Export** = full squad snapshots (casting, skills, routing, history) for sharing configured squads or publishing to agent toolboxes

**Design decision (agreed with @bradygaster):** `preset save` intentionally does NOT call `export` under the hood. They serve different workflows — presets are lightweight agent templates for bootstrapping new projects; export is a full project snapshot for portability and distribution.

**Comment by paulyuk** [comment]
Thanks @tamirdresher! Great catch 🙏

Addressed the Copilot review comment you flagged plus all 9 suggestions in [099d55eb](https://github.com/paulyuk/squad/commit/099d55eb):

1. **`SQUAD_HOME` directory validation** — `resolveSquadHome()` now throws if the path exists but is not a directory (exactly the scenario you flagged). `resolvePresetsDir()` also checks `isDirectorySync()` before returning.
2. **Path traversal protection** — Added `validateName()` that rejects preset/agent names containing `../`, path separators, or other unsafe patterns. Applied in `applyPreset()`, `savePreset()`, and per-agent during apply.
3. **Stale build cleanup** — Build step now `rmSync()`s `dist/presets/builtin` before copying, so renamed/deleted builtins don't linger.
4. **Misleading test name** — Fixed to match actual behavior.

All 24 tests passing ✅

**Comment by paulyuk** [comment]
## 🧪 E2E Test Results — Quick Start + Roaming Workflow

Ran 13 end-to-end tests against the branch using a temp `SQUAD_HOME` (clean environment, no pre-existing state). **All 13 pass** ✅

### Quick Start Flow (the README happy path)
| # | Test | Result |
|---|------|--------|
| 1 | `squad init --preset default` — auto-creates squad home, seeds built-ins, installs 5 agents | ✅ |
| 2 | Customize agents — add a custom agent manually | ✅ |
| 3 | `squad preset save my-team` — saves 6 agents (5 default + 1 custom) to SQUAD_HOME | ✅ |
| 4 | `squad init --preset my-team` in a new project — bootstraps from saved preset | ✅ |

### Preset Commands
| # | Test | Result |
|---|------|--------|
| 5 | `squad preset list` — shows default + my-team (2 presets) | ✅ |
| 6 | `squad preset show my-team` — shows 8 agents with descriptions | ✅ |
| 7 | `squad preset show default` — shows 5 built-in agents | ✅ |

### Collision Policy
| # | Test | Result |
|---|------|--------|
| 8 | Apply to existing project — all 8 agents **skipped** (no data loss) | ✅ |
| 9 | Apply with `--force` — all 8 agents **overwritten** cleanly | ✅ |

### Security / Edge Cases
| # | Test | Result |
|---|------|--------|
| 10 | Path traversal (`../../../etc`) — rejected with clear error | ✅ |
| 11 | SQUAD_HOME pointing to a **file** — throws "not a directory" error (Tamir's feedback) | ✅ |

### Roaming Simulation
| # | Test | Result |
|---|------|--------|
| 12 | `squad preset init` (local) — initializes and seeds presets | ✅ |
| 13 | Third project reuses saved preset — custom agents round-trip correctly | ✅ |

> Note: `squad preset init --remote` (GitHub repo creation/cloning) was not tested here to avoid creating real repos, but the code path is covered by the 24 unit tests.

**Tested on:** macOS, Node 22, branch `paulyuk/1038-squad-home-presets` @ 0157319a

File: packages/squad-sdk/src/resolution.ts [hunk 1]
```diff
@@ -599,3 +599,80 @@ export function resolveExternalStateDir(projectKey: string, create: boolean = tr
...
+ *
+ * @returns Absolute path to `<squad-home>/presets/`, or null if squad home
+ *          doesn't exist.
+ */
+export function resolvePresetsDir(): string | null {
+  const homeDir = resolveSquadHome();
+  if (!homeDir) return null;
+
+  const presetsDir = path.join(homeDir, 'presets');
+  if (!storage.existsSync(presetsDir)) return null;
```
[the remark this answers is not in the document]
Comment (paulyuk) [reply packages/squad-sdk/src/resolution.ts]: Fixed — resolvePresetsDir() now checks isDirectorySync() before returning.

[the remark this answers is not in the document]
Comment (paulyuk) [reply packages/squad-sdk/src/resolution.ts]: Fixed — resolvePresetsDir() now checks isDirectorySync() before returning.

File: packages/squad-sdk/src/presets/index.ts [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+  const results: PresetApplyResult[] = [];
+
+  for (const agent of manifest.agents) {
+    const sourceDir = path.join(presetAgentsDir, agent.name);
+    const destDir = path.join(targetDir, agent.name);
+
+    if (!storage.existsSync(sourceDir)) {
+      results.push({ agent: agent.name, status: 'error', reason: 'Source agent directory missing in preset' });
+      continue;
+    }
```
[the remark this answers is not in the document]
Comment (paulyuk) [reply packages/squad-sdk/src/presets/index.ts:122]: Fixed — added validateName() that rejects any name where name !== path.basename(name), blocking path traversal.

File: packages/squad-sdk/src/presets/index.ts [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+import { resolvePresetsDir, ensureSquadHome } from '../resolution.js';
+import type { PresetManifest, PresetApplyResult } from './types.js';
+
+export type { PresetManifest, PresetAgent, PresetApplyResult } from './types.js';
+
+const storage = new FSStorageProvider();
+
+function isDirSync(p: string): boolean {
+  try { return statSync(p).isDirectory(); } catch { return false; }
+}
```
[the remark this answers is not in the document]
Comment (paulyuk) [reply packages/squad-sdk/src/presets/index.ts:25]: Fixed — switched to lstatSync in isDirSync() and copyDirRecursive(). Symlinks are now skipped entirely.

File: packages/squad-cli/src/cli/commands/preset.ts [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+  let errors = 0;
+
+  for (const result of results) {
+    switch (result.status) {
+      case 'installed':
+        success(`  ✓ ${result.agent}`);
+        installed++;
+        break;
+      case 'skipped':
+        warn(`  ⊘ ${result.agent} — ${result.reason}`);
```
[the remark this answers is not in the document]
Comment (paulyuk) [reply packages/squad-cli/src/cli/commands/preset.ts]: Fixed — removed duplicate glyphs. success()/warn() handle the markers now.

File: packages/squad-sdk/package.json [hunk 5]
```diff
@@ -222,7 +226,7 @@
   ],
   "scripts": {
     "prepublishOnly": "npm run build",
-    "build": "tsc -p tsconfig.json"
+    "build": "tsc -p tsconfig.json && node -e \"const{cpSync}=require('fs');cpSync('src/presets/builtin','dist/presets/builtin',{recursive:true,filter:(s)=>!s.endsWith('.ts')})\""
```
[the remark this answers is not in the document]
Comment (paulyuk) [reply packages/squad-sdk/package.json]: Fixed — build step now rmSync()s dist/presets/builtin before cpSync.

File: test/presets.test.ts [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+  afterEach(() => {
+    if (existsSync(TMP)) rmSync(TMP, { recursive: true, force: true });
+    if (originalEnv !== undefined) {
+      process.env['SQUAD_HOME'] = originalEnv;
+    } else {
+      delete process.env['SQUAD_HOME'];
+    }
+  });
+
+  it('returns null when SQUAD_HOME is not set and ~/.squad does not exist', () => {
```
[the remark this answers is not in the document]
Comment (paulyuk) [reply test/presets.test.ts]: Fixed — test renamed to 'returns null when SQUAD_HOME points to a nonexistent path'.

File: packages/squad-sdk/src/resolution.ts [hunk 7]
```diff
@@ -599,3 +599,80 @@ export function resolveExternalStateDir(projectKey: string, create: boolean = tr
...
+    : path.join(os.homedir(), '.squad');
+
+  if (storage.existsSync(homeDir)) {
+    return homeDir;
+  }
+
+  if (create) {
+    storage.mkdirSync(homeDir, { recursive: true });
+    return homeDir;
+  }
```
[the remark this answers is not in the document]
Comment (paulyuk) [reply packages/squad-sdk/src/resolution.ts:638]: Fixed — resolveSquadHome() now throws if path exists but is not a directory.

File: packages/squad-sdk/src/presets/index.ts [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+  const presetsDir = resolvePresetsDir();
+  if (!presetsDir) {
+    return [{ agent: presetName, status: 'error', reason: 'No presets directory found. Run `squad preset init` first.' }];
+  }
+
+  const presetDir = path.join(presetsDir, presetName);
+  const manifest = loadPresetManifest(presetDir);
+  if (!manifest) {
+    return [{ agent: presetName, status: 'error', reason: `Preset '${presetName}' not found` }];
+  }
```
[the remark this answers is not in the document]
Comment (paulyuk) [reply packages/squad-sdk/src/presets/index.ts:105]: Fixed — applyPreset() validates presetName with validateName() before path.join.

File: packages/squad-sdk/src/presets/index.ts [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+      results.push({ agent: agent.name, status: 'skipped', reason: 'Already exists (use --force to overwrite)' });
+      continue;
+    }
+
+    try {
+      copyDirRecursive(sourceDir, destDir);
+      results.push({ agent: agent.name, status: 'installed' });
+    } catch (err) {
+      results.push({ agent: agent.name, status: 'error', reason: String(err) });
+    }
```
[the remark this answers is not in the document]
Comment (paulyuk) [reply packages/squad-sdk/src/presets/index.ts:138]: Fixed — when --force is set, destDir is rmSync'd before copyDirRecursive.

File: packages/squad-cli/src/cli/commands/preset.ts [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+  const presetsDir = path.join(homeDir, 'presets');
+
+  const seeded = seedBuiltinPresets();
+
+  success('Presets directory initialized');
+  info(`  Path: ${presetsDir}`);
+  if (seeded.length > 0) {
+    info(`  Built-in presets installed: ${seeded.join(', ')}`);
+  }
+  info(`  Run 'squad preset list' to see available presets.`);
```
Comment (bradygaster) [inline packages/squad-cli/src/cli/commands/preset.ts:104]: so this and `init` related features don't collide with the `--roles` switch, you think?

File: packages/squad-cli/src/cli/commands/preset.ts [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+      execSync(`gh repo view ${repoFullName} --json name`, { stdio: 'pipe' });
+      repoExists = true;
+    } catch {
+      repoExists = false;
+    }
+
+    if (!repoExists) {
+      info(`Creating private repo ${repoFullName}...`);
+      try {
+        execSync(`gh repo create ${repoName} --private --source "${homeDir}" --push --description "Squad home — presets and config"`, {
```
Comment (bradygaster) [inline packages/squad-cli/src/cli/commands/preset.ts:210]: i presume we don't want a git-provider unspecific way?

File: packages/squad-cli/src/cli/commands/preset.ts [hunk 12]
```diff
@@ -0,0 +1,389 @@
...
+// Subcommand: list
+// ============================================================================
+
+async function presetList(): Promise<void> {
+  const presetsDir = resolvePresetsDir();
+
+  if (!presetsDir) {
+    info('No presets directory found.');
+    info('  Run `squad preset init --remote` to set up with a GitHub repo (recommended).');
+    info('  Or `squad preset init` for local-only setup.');
```
Comment (bradygaster) [inline packages/squad-cli/src/cli/commands/preset.ts:258]: maybe add `--no-workflows` here, too

File: packages/squad-sdk/src/presets/builtin/default/agents/devrel/charter.md [hunk 13]
```diff
@@ -0,0 +1,30 @@
+# devrel — Developer Relations
+
+> I make your project approachable — if a developer can't get started in 5 minutes, that's on me.
```
Comment (bradygaster) [inline packages/squad-sdk/src/presets/builtin/default/agents/devrel/charter.md:3]: nice rule

**Review by bradygaster — APPROVED** [review]
no big concerns but flagging some stuff. ;)

**Comment by bradygaster** [comment]
thanks @paulyuk !

File: packages/squad-cli/src/cli/commands/preset.ts (hunk 10 above)
Comment (paulyuk) [reply packages/squad-cli/src/cli/commands/preset.ts:104]: I checked all good.

File: packages/squad-cli/src/cli/commands/preset.ts (hunk 12 above)
Comment (paulyuk) [reply packages/squad-cli/src/cli/commands/preset.ts:258]: making a new issue to track.
